src/tools/compiletest/src/runtest.rs

```diff
@@ -2325,15 +2341,100 @@ impl<'test> TestCx<'test> {
             let res = self.cmd2procres(
                 Command::new(&self.config.docck_python)
                     .arg(root.join("src/etc/htmldocck.py"))
-                    .arg(out_dir)
+                    .arg(&out_dir)
                     .arg(&self.testpaths.file),
             );
             if !res.status.success() {
-                self.fatal_proc_rec("htmldocck failed!", &res);
+                self.fatal_proc_rec_with_ctx("htmldocck failed!", &res, |mut this| {
+                    this.compare_to_default_rustdoc(&out_dir)
+                });
             }
         }
     }
 
+    fn compare_to_default_rustdoc(&mut self, out_dir: &Path) {
+        println!("info: generating a diff against nightly rustdoc");
+
+        let suffix =
+            self.safe_revision().map_or("nightly".into(), |path| path.to_owned() + "-nightly");
+        let compare_dir = output_base_dir(self.config, self.testpaths, Some(&suffix));
+        let _ = fs::remove_dir_all(&compare_dir);
+        create_dir_all(&compare_dir).unwrap();
+
+        // We need to create a new struct for the lifetimes on `config` to work.
+        let new_rustdoc = TestCx {
+            config: &Config {
+                // FIXME: use beta or a user-specified rustdoc instead of
+                // hardcoding the default toolchain
+                rustdoc_path: Some("rustdoc".into()),
+                // Needed for building auxiliary docs below
+                rustc_path: "rustc".into(),
+                ..self.config.clone()
+            },
+            ..*self
+        };
+
+        let output_file = TargetLocation::ThisDirectory(new_rustdoc.aux_output_dir_name());
+        let mut rustc = new_rustdoc.make_compile_args(
+            &new_rustdoc.testpaths.file,
+            output_file,
+            EmitMetadata::No,
+            AllowUnused::Yes,
+        );
+        rustc.arg("-L").arg(&new_rustdoc.aux_output_dir_name());
+        new_rustdoc.build_all_auxiliary(&mut rustc);
+
+        let proc_res = new_rustdoc.document(&compare_dir);
+        if !proc_res.status.success() {
+            proc_res.fatal(Some("failed to run nightly rustdoc"), || ());
+        }
+
+        // NOTE: this is fine since compiletest never runs out-of-tree
+        let tidy = concat!(env!("CARGO_MANIFEST_DIR"), "/tidy-rustdoc.sh");
+        // FIXME: this overwrites `out_dir` in place, maybe we should make a copy?
+        let status = Command::new(tidy)
+            .arg(out_dir)
+            .spawn()
+            .expect("tidy-rustdoc not found")
+            .wait()
+            .unwrap();
+        if !status.success() {
+            self.fatal("failed to run tidy - is it installed?");
+        }
+        let status = Command::new(tidy).arg(&compare_dir).spawn().unwrap().wait().unwrap();
+        if !status.success() {
+            self.fatal("failed to run tidy");
+        }
+
+        let has_delta = Command::new("delta")
+            .arg("--version")
```

Comment:
Regarding the possible confusion with other executables named "delta", would one possibility be to call delta with a command line argument that only delta would have, in addition to `--version`? I admit it might look a bit odd, but e.g. `delta --version --show-config` also just prints the version number but would be very likely to fail for anything that's not the intended delta.